types/header.go

```diff
@@ -0,0 +1,239 @@
+	return strings.ToUpper(hex.EncodeToString(h))
+}
+
+func (h Hash) MarshalJSON() ([]byte, error) {
+	s := strings.ToUpper(hex.EncodeToString(h))
+	jbz := make([]byte, len(s)+2)
+	jbz[0] = '"'
+	copy(jbz[1:], s)
+	jbz[len(jbz)-1] = '"'
+	return jbz, nil
+}
+
+func (h *Hash) UnmarshalJSON(data []byte) error {
+	if len(data) < 2 || data[0] != '"' || data[len(data)-1] != '"' {
+		return fmt.Errorf("invalid hex string: %s", data)
+	}
+	bz2, err := hex.DecodeString(string(data[1 : len(data)-1]))
+	if err != nil {
+		return err
+	}
+	*h = bz2
+	return nil
+}
+
+// Header abstracts all methods required to perform header sync.
+type H interface {
+	// New creates new instance of a header.
+	New() H
+	// Hash returns hash of a header.
+	Hash() [32]byte
+	// Number returns the height of a header.
+	Number() int64
+	// LastHeader returns the hash of last header before this header (aka. previous header hash).
+	LastHeader() [32]byte
+	// Timestamp returns time when header was created.
+	Timestamp() time.Time
+	// IsRecent checks if header is recent against the given blockTime.
+	IsRecent(duration time.Duration) bool
+	// IsExpired checks if header is expired against trusting period.
+	IsExpired() bool
+	// VerifyAdjacent validates adjacent untrusted header against trusted header.
+	VerifyAdjacent(H) error
+	// VerifyNonAdjacent validates non-adjacent untrusted header against trusted header.
+	VerifyNonAdjacent(H) error
+	// Verify performs basic verification of untrusted header.
+	Verify(H) error
+	// Validate performs basic validation to check for missed/incorrect fields.
+	Validate() error
+
+	encoding.BinaryMarshaler
+	encoding.BinaryUnmarshaler
+}
+
+var _ H = (*Header)(nil)
+
+// New creates new instance of a header.
+func (h *Header) New() H {
+	return &Header{}
+}
+
+// Number returns the height of a header.
+func (h *Header) Number() int64 {
+	return int64(h.Height)
+}
+
+// Timestamp returns time when header was created.
+func (h *Header) Timestamp() time.Time {
+	return time.Unix(int64(h.Time), 0)
+}
+
+// LastHeader returns the hash of last header before this header (aka. previous header hash).
+func (h *Header) LastHeader() [32]byte {
+	return h.LastHeaderHash
+}
+
+// IsRecent checks if header is recent against the given blockTime.
+func (h *Header) IsRecent(blockTime time.Duration) bool {
+	return time.Since(time.Unix(int64(h.Time), 0)) <= blockTime
+}
+
+// TODO(@Wondertan): We should request TrustingPeriod from the network's state params or
+//  listen for network params changes to always have a topical value.
+
+// TrustingPeriod is period through which we can trust a header's validators set.
+//
+// Should be significantly less than the unbonding period (e.g. unbonding
+// period = 3 weeks, trusting period = 2 weeks).
+//
+// More specifically, trusting period + time needed to check headers + time
+// needed to report and punish misbehavior should be less than the unbonding
+// period.
+var TrustingPeriod = 168 * time.Hour
```

Comment:
[non-blocking] Comment Rollkit might have a different shorter or longer trusting period than node. The fraud proof window time might depend on this as well, if you are an optimistic Rollup. This should be toggleable in the future and the default value should be a deliberate value.